File: src/main/java/tmi/recipe/RecipeItemManager.java
package tmi.recipe;

import arc.Core;
import arc.func.Boolf;
import arc.func.Func;
import arc.graphics.g2d.TextureRegion;
import arc.struct.ObjectIntMap;
import arc.struct.ObjectMap;
import arc.struct.OrderedMap;
import arc.struct.Seq;
import mindustry.Vars;
import mindustry.ctype.ContentType;
import mindustry.ctype.UnlockableContent;
import tmi.recipe.types.RecipeItem;

@SuppressWarnings({"unchecked", "rawtypes"})
public class RecipeItemManager {
  protected static final RecipeItem ERROR = new RecipeItem<>("error") {
    @Override public int ordinal() {return -1;}
    @Override public int typeID() {return -1;}
    @Override public String name() {return "<error>";}
    @Override public String localizedName() {return "<error>";}
    @Override public TextureRegion icon() {return Core.atlas.find("error");}
    @Override public boolean hidden() {return true;}
  };

  protected static final OrderedMap<Boolf<Object>, Func<?, RecipeItem<?>>> wrapper = new OrderedMap<>();

  protected final ObjectMap<Object, RecipeItem<?>> recipeItems = new ObjectMap<>();
  protected final ObjectMap<String, RecipeItem<?>> itemNameMap = new ObjectMap<>();

  public static <T> void registerWrapper(Boolf<Object> matcher, Func<T, RecipeItem<T>> factory){
    wrapper.put(matcher, (Func) factory);
  }

  static {
    registerWrapper(e -> e instanceof UnlockableContent, RecipeUnlockableContent::new);
  }

  public <T, R extends RecipeItem<T>> R addItemWrap(T item, R recipeItem){
    recipeItems.put(item, recipeItem);
    itemNameMap.put(recipeItem.name(), recipeItem);

    return recipeItem;
  }

  public <T> RecipeItem<T> getItem(T item) {
    return (RecipeItem<T>) recipeItems.get(item, () -> {
      for (ObjectMap.Entry<Boolf<Object>, Func<?, RecipeItem<?>>> entry : wrapper) {
        if (entry.key.get(item)) {
          RecipeItem<?> res = (RecipeItem<?>) ((Func) entry.value).get(item);
          itemNameMap.put(res.name(), res);
          return res;
        }
      }
      return ERROR;
    });
  }

  public <T> RecipeItem<T> getByName(String name){
    return (RecipeItem<T>) itemNameMap.get(name, ERROR);
  }

  public Seq<RecipeItem<?>> getList(){
    return recipeItems.values().toSeq().sort();
  }

  protected static class RecipeUnlockableContent extends RecipeItem<UnlockableContent>{
    private static final ObjectIntMap<ContentType> mirror = new ObjectIntMap<>();

    static {
      mirror.put(ContentType.item, 0);
      mirror.put(ContentType.liquid, 1);
      mirror.put(ContentType.block, 2);
      mirror.put(ContentType.unit, 3);
    }

    public RecipeUnlockableContent(UnlockableContent item) {
      super(item);
    }

    @Override public int ordinal() {return item.id;}
    @Override public int typeID() {return mirror.get(item.getContentType(), item.getContentType().ordinal());}
    @Override public String name() {return item.name;}
    @Override public String localizedName() {return item.localizedName;}
    @Override public TextureRegion icon() {return item.uiIcon;}
    @Override public boolean hidden() {return false;}
    @Override public boolean locked() {return !item.unlockedNow();}
    @Override public boolean hasDetails() {return true;}
    @Override public void displayDetails() {Vars.ui.content.show(item);}
  }
}

File: src/main/kotlin/tmi/recipe/RecipeItemManager.kt
package tmi.recipe

import arc.Core
import arc.func.Boolf
import arc.func.Func
import arc.graphics.g2d.TextureRegion
import arc.struct.ObjectIntMap
import arc.struct.ObjectMap
import arc.struct.OrderedMap
import arc.struct.Seq
import mindustry.Vars
import mindustry.ctype.ContentType
import mindustry.ctype.UnlockableContent
import mindustry.world.Block
import tmi.util.invoke
import tmi.recipe.types.RecipeItem
import tmi.recipe.types.SingleItemMark
import tmi.util.Consts
import tmi.util.set

class RecipeItemManager {
  private val recipeItems = ObjectMap<Any, RecipeItem<*>>()
  private val itemNameMap = ObjectMap<String, RecipeItem<*>>()

  val wrapper = OrderedMap<Boolf<Any>, Func<*, RecipeItem<*>>>()

  init {
    registerWrapper<UnlockableContent> { RecipeUnlockableContent(it) }
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> registerWrapper(matcher: Boolf<Any>, factory: Func<T, RecipeItem<T>>) {
    wrapper[matcher] = factory as Func<*, RecipeItem<*>>
  }

  @Suppress("UNCHECKED_CAST")
  inline fun <reified T> registerWrapper(factory: Func<T, RecipeItem<T>>) {
    wrapper[Boolf{ e: Any? -> e is T }] = factory as Func<*, RecipeItem<*>>
  }

  fun <T, R : RecipeItem<T>> addItemWrap(item: T, recipeItem: R): R {
    recipeItems.put(item, recipeItem)
    itemNameMap.put(recipeItem.name, recipeItem)

    return recipeItem
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> getItem(item: T): RecipeItem<T> {
    return recipeItems.get(item) {
      for (entry in wrapper) {
        if (entry.key[item]) {
          val res = (entry.value as Func<T, RecipeItem<T>>)(item)
          itemNameMap.put(res.name, res)
          return@get res
        }
      }
      ERROR
    } as RecipeItem<T>
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> getByName(name: String): RecipeItem<T> {
    return itemNameMap[name, ERROR] as RecipeItem<T>
  }

  val list: Seq<RecipeItem<*>>
    get() = recipeItems.values().toSeq().sort()

  private class RecipeUnlockableContent(item: UnlockableContent) : RecipeItem<UnlockableContent>(item) {
    override val ordinal = item.id.toInt()
    override val typeTag: String by lazy {
      if (item is Block) Core.bundle["database-tag." + item.databaseTag, item.databaseTag]
      else Core.bundle["database-category." + item.databaseCategory, item.databaseCategory]
    }
    override val typeID: Int = item.contentType.ordinal
    override val ownMod: String = item.minfo.mod?.name?: Consts.VANILLA
    override val name: String = item.name
    override val localizedName: String = item.localizedName
    override val icon: TextureRegion = item.uiIcon?:throw IllegalStateException("Item $name no icon")
    override val hidden = false
    override val locked = !item.unlockedNow()
    override val hasDetails = true

    override fun displayDetails() = Vars.ui.content.show(item)

    companion object {
      private val mirror = ObjectIntMap<ContentType>()

      init {
        mirror.put(ContentType.item, 0)
        mirror.put(ContentType.liquid, 1)
        mirror.put(ContentType.block, 2)
        mirror.put(ContentType.unit, 3)
      }
    }
  }

  companion object {
    private val ERROR = object : SingleItemMark("<error>") {
      override val ownMod: String get() = "<error>"
      override val icon get() = Core.atlas.find("error")
      override val hidden = true
    }
  }
}
